core/src/main/java/com/linecorp/armeria/server/file/FileService.java

```diff
@@ -385,6 +369,58 @@ private HttpFile findFile(ServiceRequestContext ctx, HttpRequest req) {
         });
     }
 
+    private CompletableFuture<@Nullable HttpFile> findFileWithIndexPath(
+            ServiceRequestContext ctx, String decodedMappedPath,
+            Set<ContentEncoding> encodings, boolean decompress) {
+        // Redirect to the slash appended path if:
+        // 1) /index.html exists or
+        // 2) it has a directory listing.
+        final String indexPath = decodedMappedPath + "/index.html";
+        return findFile(ctx, indexPath, encodings, decompress).thenCompose(indexFile -> {
+            if (indexFile != null) {
+                return UnmodifiableFuture.completedFuture(true);
+            }
+
+            if (!config.autoIndex()) {
+                return UnmodifiableFuture.completedFuture(false);
+            }
+
+            return config.vfs().canList(ctx.blockingTaskExecutor(), decodedMappedPath);
+        }).thenApply(canList -> {
+            if (canList) {
+                try (TemporaryThreadLocals ttl = TemporaryThreadLocals.acquire()) {
+                    final StringBuilder locationBuilder = ttl.stringBuilder()
+                                                             .append(ctx.path())
+                                                             .append('/');
+                    if (ctx.query() != null) {
+                        locationBuilder.append('?')
+                                       .append(ctx.query());
+                    }
+                    return HttpFile.ofRedirect(locationBuilder.toString());
+                }
+            } else {
+                return HttpFile.nonExistent();
+            }
+        });
+    }
+
+    private CompletableFuture<@Nullable HttpFile> findFileWithExtensions(
+            ServiceRequestContext ctx, @Nullable Iterator<String> extensionIterator, String path,
+            Set<ContentEncoding> supportedEncodings, boolean decompress) {
+        if (extensionIterator == null || !extensionIterator.hasNext()) {
+            return UnmodifiableFuture.completedFuture(null);
+        }
+
+        final String extension = extensionIterator.next();
+        return findFile(ctx, path + '.' + extension, supportedEncodings, decompress).thenCompose(file -> {
+            if (file != null) {
+                return UnmodifiableFuture.completedFuture(file);
```

Comment:
Question) I'm not sure if we should be redirecting to the location with the extension like done in the other location, or we can just return the file as-is.

I think either way should work, but asking just in case I'm missing anything.

> I prefer clean URL patterns without a trailing slash so /app/projects is preferred over /app/projects/. 

Also, if it isn't necessary to redirect then I thought maybe this is solvable in a simpler way

Reply:
Talked internally, and it seems like other web servers are also employing the redirect logic. I think the current approach is probably fine then 👍